### PR TITLE
Directed SPDP messages are sent via multicast

### PR DESCRIPTION
**Pull request by jrw972**
Problem
-------

In some environments, unicast is scheduled before multicast.  For
completing authentication, this causes a delay since the SPDP message
that is necessary to start authentication on a remote may be delivered
after the authentication messages (even though the code clearly sends
them in the correct order).

Solution
--------

Send directed SPDP messages via unicast.

### CODE REVIEW

File: dds/DCPS/RTPS/Spdp.cpp [hunk 1]
```diff
@@ -2161,8 +2170,9 @@ bool Spdp::announce_domain_participant_qos()
 }
 
 #if !defined _MSC_VER || _MSC_VER > 1700
-const Spdp::SpdpTransport::WriteFlags Spdp::SpdpTransport::SEND_TO_LOCAL;
-const Spdp::SpdpTransport::WriteFlags Spdp::SpdpTransport::SEND_TO_RELAY;
+const Spdp::SpdpTransport::WriteFlags Spdp::SpdpTransport::SEND_MULTICAST;
+const Spdp::SpdpTransport::WriteFlags Spdp::SpdpTransport::SEND_RELAY;
+ const Spdp::SpdpTransport::WriteFlags Spdp::SpdpTransport::SEND_DIRECT;
```
Comment (mitza-oci) [inline dds/DCPS/RTPS/Spdp.cpp:2175]:
```suggestion
const Spdp::SpdpTransport::WriteFlags Spdp::SpdpTransport::SEND_DIRECT;
```

File: dds/DCPS/RTPS/Spdp.cpp [hunk 2]
```diff
@@ -2664,24 +2674,29 @@ Spdp::SpdpTransport::write_i(const DCPS::RepoId& guid, WriteFlags flags)
...
 
-  if ((flags & SEND_TO_LOCAL) && !outer->config_->rtps_relay_only()) {
+  if ((flags & SEND_MULTICAST) && !outer->config_->rtps_relay_only()) {
     typedef OPENDDS_SET(ACE_INET_Addr)::const_iterator iter_t;
     for (iter_t iter = send_addrs_.begin(); iter != send_addrs_.end(); ++iter) {
       send(*iter);
     }
   }
 
+  if (((flags & SEND_DIRECT) || !outer->config_->rtps_relay_only()) &&
```
Comment (mitza-oci) [inline dds/DCPS/RTPS/Spdp.cpp:2693]: Why is this an `||`?  If the command is to send it "directly" and the configuration is "relay only," should that prevent this `send()`?

**Review by mitza-oci — CHANGES_REQUESTED** [review]

**Review by mitza-oci — APPROVED** [review]

File: dds/DCPS/RTPS/Spdp.cpp [hunk 3]
```diff
@@ -680,8 +683,8 @@ Spdp::handle_participant_data(DCPS::MessageId id,
     // Since we've just seen a new participant, let's send out our
     // own announcement, so they don't have to wait.
     if (from != ACE_INET_Addr()) {
-      if (from_relay) {
-        tport_->write_i(guid, SpdpTransport::SEND_TO_RELAY);
+      if (!from_relay && from_relay) {
```
Comment (iguessthislldo) [inline dds/DCPS/RTPS/Spdp.cpp:686]: Wouldn't this always be false?

**Review by iguessthislldo — CHANGES_REQUESTED** [review]

File: dds/DCPS/RTPS/Spdp.cpp (hunk 3 above)
Comment (jrw972) [reply dds/DCPS/RTPS/Spdp.cpp]: Good catch.  That probably saved me hours of debugging.

**Pull request merged**